File: rs-typeproto/src/compiled_file.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::ast::Breadcrumbs;
use crate::instructions::Instruction;
use crate::types::CanonicalLangType;
use crate::vm::RuntimeBadValue;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompiledFile {
    pub path: String,
    pub types: HashMap<String, CanonicalLangType>,
    pub chunks: Vec<InstructionChunk>,
    pub exports: HashMap<String, CompiledExport>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct InstructionChunk {
    pub constant_table: Vec<CompiledConstant>,
    pub instructions: Vec<Instruction>,
}

impl InstructionChunk {
    pub fn add_constant(&mut self, constant: CompiledConstant) -> usize {
        let existing_index =
            self.constant_table
                .iter()
                .enumerate()
                .find_map(|(i, existing_constant)| {
                    if existing_constant == &constant {
                        Some(i)
                    } else {
                        None
                    }
                });

        if let Some(existing_index) = existing_index {
            existing_index
        } else {
            self.constant_table.push(constant);
            self.constant_table.len() - 1
        }
    }

    pub fn write_instruction(&mut self, instruction: Instruction) {
        self.instructions.push(instruction);
    }

    pub fn write_literal(&mut self, constant: CompiledConstant) {
        let index = self.add_constant(constant);
        self.write_instruction(Instruction::Literal(index));
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum CompiledConstant {
    String(String),
    Integer(i64),
    Float(f64),
    Error(RuntimeBadValue),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum CompiledExport {
    Type(String),
    Chunk(usize),
    Constant(CompiledConstant),
}

File: ktcause/src/main/kotlin/CompiledFile.kt
import com.dallonf.ktcause.Instruction
import com.dallonf.ktcause.ResolvedFile
import com.dallonf.ktcause.Resolver
import com.dallonf.ktcause.ast.Breadcrumbs
import com.dallonf.ktcause.types.*

data class CompiledFile(
    val path: String,
    val types: Map<CanonicalLangTypeId, CanonicalLangType>,
    val chunks: List<InstructionChunk>,
    val exports: Map<String, CompiledExport>,
    // TODO: I don't love that we need this for error reporting;
    // ideally a CompiledFile would be a standalone artifact distinct from the AST and resolver
    val resolved: ResolvedFile?
) {
    fun toFileDescriptor(): Resolver.ExternalFileDescriptor {
        val exportDescriptors = mutableMapOf<String, ValueLangType>()
        for ((exportName, export) in exports) {
            when (export) {
                is CompiledExport.Type -> {
                    val canonicalType = types[export.typeId]
                    requireNotNull(canonicalType) { "$path describes a type (${export.typeId}) but doesn't define it" }
                    exportDescriptors[exportName] = canonicalType
                }

                is CompiledExport.Function -> TODO()
                is CompiledExport.Value -> TODO()
            }
        }

        return Resolver.ExternalFileDescriptor(exportDescriptors)
    }

    data class InstructionChunk(val constantTable: List<CompiledConstant>, val instructions: List<Instruction>)

    data class MutableInstructionChunk(
        val constantTable: MutableList<CompiledConstant> = mutableListOf(),
        val instructions: MutableList<Instruction> = mutableListOf(),
    ) {
        fun toInstructionChunk() = InstructionChunk(constantTable.toList(), instructions.toList())

        fun addConstant(constant: CompiledConstant): Int {
            val existingIndex = constantTable.asSequence().withIndex()
                .firstNotNullOfOrNull { (i, existing) -> if (existing == constant) i else null }

            return if (existingIndex != null) {
                existingIndex
            } else {
                constantTable.add(constant)
                constantTable.lastIndex
            }
        }

        fun writeInstruction(instruction: Instruction) {
            instructions.add(instruction)
        }

        fun writeLiteral(constant: CompiledConstant) {
            val index = addConstant(constant)
            writeInstruction(Instruction.Literal(index))
        }
    }

    sealed interface CompiledConstant {
        data class StringConst(val value: String) : CompiledConstant
        data class IntegerConst(val value: Long) : CompiledConstant
        data class FloatConst(val value: Double) : CompiledConstant
        data class ErrorConst(val filePath: String, val breadcrumbs: Breadcrumbs, val error: ErrorValueLangType) :
            CompiledConstant
    }

    sealed interface CompiledExport {
        data class Type(val typeId: CanonicalLangTypeId) : CompiledExport
        data class Function(val chunkIndex: Int, val type: ValueLangType) : CompiledExport
        data class Value(val constant: CompiledConstant, val type: ValueLangType) : CompiledExport
    }
}